junit5/src/main/java/org/jboss/weld/junit5/WeldJunit5Extension.java

```diff
@@ -196,85 +198,106 @@ private boolean methodRequiresExplicitParamInjection(ParameterContext pc) {
-                            AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
-                                field.setAccessible(true);
-                                return null;
-                            });
-                            fieldInstance = field.get(testInstance);
-                        }
-                        if (fieldInstance != null && fieldInstance instanceof WeldInitiator) {
-                            initiator = (WeldInitiator) fieldInstance;
-                            foundInitiatorFields.add(field);
-                        } else {
-                            // Field with other type than WeldInitiator was annotated with @WeldSetup
-                            throw new IllegalStateException("@WeldSetup annotation should only be used on a field of type"
-                                    + " WeldInitiator but was found on a field of type " + field.getType() + " which is declared "
-                                    + "in class " + field.getDeclaringClass());
-                        }
-                    }
-                }
-            }
-            // Multiple occurrences of @WeldSetup in the hierarchy will lead to an exception
-            if (foundInitiatorFields.size() > 1) {
-                throw new IllegalStateException(foundInitiatorFields.stream().map(f -> "Field type - " + f.getType() + " which is "
-                        + "in " + f.getDeclaringClass()).collect(Collectors.joining("\n", "Multiple @WeldSetup annotated fields found, "
-                        + "only one is allowed! Fields found:\n", "")));
-            }
+            // iterate through the testInstance, the enclosing instance (in case of nested tests),
+            // the enclosing instance of the enclosing instance (in cases of twice nested tests) and so on
+            // until we find a WeldInitiator
+            final List<Object> allTestInstances = new ArrayList<>(context.getRequiredTestInstances().getAllInstances());
+            Collections.reverse(allTestInstances); // so we can iterate from inner-most to outer-most
+            WeldInitiator initiator = allTestInstances.stream()
+                    .map(this::findInitiatorInInstance)
+                    .filter(Objects::nonNull)
+                    .findFirst()
+                    .orElseGet(() -> getDefaultInitiator(context, testInstance));
+            setInitiatorToStore(context, initiator);
 
-            // at this point we can be sure that either no or exactly one WeldInitiator was found
-            if (initiator == null) {
-                Weld weld = WeldInitiator.createWeld();
-                WeldInitiator.Builder builder = WeldInitiator.from(weld);
+            // this ensures the test class is injected into
+            // in case of nested tests, this also injects into any outer classes
+            initiator.addObjectsToInjectInto(new HashSet<>(allTestInstances));
```

Comment:
I prefer the first option. It would lead to an reasonably explicit and somewhat useful exception that tells the programmer exactly that the problem is with the outer injection points. With a bit of effort one could even detect this case before-hand and print out a more explicit explanation of what's going on. The second option would be very non-intuitive in certain cases: Adding a nested class with an empty test method and a empty WeldIntiator would lead to a failing test if any injection point in the outer class is used in a `@BeforeEach` or `@AfterEach` method of the outer class, because those also get executed for test methods in inner test classes. And the error would (most likely) be a relatively unhelpful NullPointerException.

Reply:
I have expanded the readme accordingly.